### PR TITLE
Support pure `pyproject-toml` packages

### PR DESCRIPTION
**Pull request by scbedd**
See title 👍 .

Of course, this is purely "support" and we'll probably find _some_ warts when we migrate the first packages. This has been a long asked-for feature that I've just not had a day to dedicate to. Had a good confluence of availability and lack of interruptions this past Friday.
 
Working through test failures. With the advent of `setuptools>=74.1` extension packages are now supported directly from `setuptools` 

Todo:

- [x] Verify unbroken [nightly](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=4252346&view=results) etc
- [x] Fix outstanding build failures
- [x] Try using `setuptools.meta` one last time, I have a feeling that patches have resolved the issues I was seeing this past jan
- [x] Finish up tests
- [x] Upgrade setuptools for the repo to >= `74.1.3` to prepare for pure pyproject.toml extension packages

### CODE REVIEW

**Comment by scbedd** [comment]
@mccoyp this is ready to rock and roll.

File: tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py [hunk 1]
```diff
@@ -320,6 +326,172 @@ def setup(*args, **kwargs):
...
+        return setup_filename
+
+    return None
+
+
+def parse_setup(
+    setup_filename_or_folder: str,
+):
+    """
+    Used to evaluate a setup.py (or a directory containing a setup.py) and return a tuple containing:
```
Comment (mccoyp) [inline tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py:433]:
```suggestion
    Used to evaluate a pyproject.toml or setup.py (or a directory containing either) and return a tuple containing:
```

File: tools/azure-sdk-tools/tests/integration/scenarios/pyproject_build_config/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+The `pyproject.toml` and `setup.py` files in this folder are used to verify successful azure-sdk build configuration from a parsed setup.py.
```
Comment (mccoyp) [inline tools/azure-sdk-tools/tests/integration/scenarios/pyproject_build_config/README.md:1]:
```suggestion
The `pyproject.toml` and `setup.py` files in this folder are used to verify successful azure-sdk build configuration from a parsed configuration file.
```

File: tools/azure-sdk-tools/tests/integration/scenarios/pyproject_project_def_with_extension/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+requires = ["setuptools>=61.0.0", "wheel"]  # Requires 61.0.0 for dynamic version
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "azure-keyvault-keys"
+authors = [
+    {name = "Microsoft Corporation", email = "azurekeyvault@microsoft.com"}
+    ]
+description = "Microsoft Azure Key Vault Keys Client Library for Python"
+requires-python = ">=3.7"
```
Comment (mccoyp) [inline tools/azure-sdk-tools/tests/integration/scenarios/pyproject_project_def_with_extension/pyproject.toml:11]: nit:
```suggestion
requires-python = ">=3.8"
```

File: tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py [hunk 4]
```diff
@@ -320,6 +326,172 @@ def setup(*args, **kwargs):
...
+    project_config = toml_dict.get("project", None)
+
+    # to pull a version from pyproject.toml, we need to get a dynamic version out. We can ask
+    # setuptools to give us the metadata for a package, but that will involve _partially building_ the package
+    # to create an egginfo folder. This is a very expensive operation goes against the entire point of
+    # "give me the package metadata for this folder."
+    # We can avoid this expensive operation if we parse the version out of the _version or version file directly.
+    parsed_version = project_config.get("version", None)
+    if not parsed_version:
+        parsed_version_py = get_version_py(pyproject_filename)
```
[the remark this answers is not in the document]
Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py:344]: Yeah, good thought.

File: tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py [hunk 5]
```diff
@@ -320,6 +326,172 @@ def setup(*args, **kwargs):
     )
     # fmt: on
 
+
+def parse_pyproject(
+    pyproject_filename: str,
+) -> Tuple[str, str, str, List[str], bool, str, str, Dict[str, Any], bool, List[str], List[str], str, List[Extension]]:
```
[the remark this answers is not in the document]
Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/parsing/parse_functions.py:332]: Yeah I'll do a pass on the docstring refs.

File: tools/azure-sdk-tools/tests/integration/scenarios/pyproject_project_def_with_extension/pyproject.toml (hunk 3 above)
Comment (scbedd) [reply tools/azure-sdk-tools/tests/integration/scenarios/pyproject_project_def_with_extension/pyproject.toml:11]: Keepin me honest :D

**Comment by scbedd** [comment]
@mccoyp addressed your feedback + updated a couple comments.

**Review by mccoyp — APPROVED** [review]

**Pull request merged**